### ISSUE TITLE
Improving Automation of Matching CRD Properties to Kubernetes Core Resources

### ISSUE THREAD

**Issue by tylergu** [xlab-uiuc/acto#296]
## Problem
@MarkintoshZ found out that the automatic schema matching for the strimzi-kafka-operator is not working well. Although Acto supports specifying the mapping through manual annotation, the process is very repetitive and tricky.

The current implementation for automatically matching CRD properties to Kubernetes core resources has some major limitations.
Here are the two main reasons causing this problem:
1. Currently, the semantic testcase generation for Kubernetes core resources and matching properties to Kubernetes core resources are intertwined together. We **manually** write the schema for Kubernetes core resources and their semantic testcases under the same class `K8sSchema`. Since this process is manual and on-demand(we only write semantic testcases for a small subset of Kubernetes core resources), the schemas are incomplete.

   The right design is to separate the schema matching and semantic test generation. Schema matching is very easy to be automated and become complete, while the semantic test generation does not have to be complete and can be added on-demand.
2. Another challenge is that the Kubernetes core resource schema frequently changes from version to version. This makes manually writing schema even harder: we would need to manually write schemas for several Kubernetes versions that we want to support.

## Solution
The solution is to separate the schema matching and semantic test generation. The schema matching can be automated by parsing the Kubernetes core resource schemas. I drafted an implementation to use Python's `inspect` module to extract the schemas from the Kubernetes Python Client module, this way we can get the schemas of Kubernetes core resources for the specific Kubernetes Python Client version. https://github.com/xlab-uiuc/acto/commit/d171bd8988e2b91a9a609f0bb5d8098093f68179

However, this approach has some drawbacks: to support matching for different Kubernetes version, users would need to install and link a different version of Kubernetes Python Client. This process is unnecessarily difficult.

A better approach is to directly download and parse Kubernetes API's swagger spec. This allows users to write the Kubernetes version in a configuration, and Acto can download the corresponding swagger spec and do the matching.

### Other challenges
Different controller toolchain generates the CRD slightly differently. We would need to take the MetaClass approach to loose the matching logic for some schemas.

**Comment by tianyin** [comment]
This is very cool, especially if the goal is to support different operators. I would imagine it will be the case in 523 next semester.

**Comment by tianyin** [comment]
> A better approach is to directly download and parse Kubernetes API's swagger spec. This allows users to write the Kubernetes version in a configuration, and Acto can download the corresponding swagger spec and do the matching.

We should create a task for that.

**Comment by MarkintoshZ** [comment]
Discussed this issue with @tylergu last night, and we share a similar high-level perspective on refactoring. After evaluating various use cases, we agreed to adopt the design of implementing generators as annotated functions. This approach prioritizes function composition over inheritance, resulting in shorter and more understandable value generator code. This change is expected to benefit new Acto users in particular.

# Refactoring Plans

### Matching Use Cases
- Matching a superset of a known schema properties
- Matching a subset of a known schema properties
- Matching standalone primitive schema (e.g. `replicas`)

### Generator Use Cases
- Generate both testcases from both standalone and composite schemas
- Implementing a generator for a schema requires knowledge of the schema (type and name of properties). Is it possible to it with strong typing?
- How to implement a generator for a schema that is a superset or subset of a known schema?

### Solution
- Have a two stage matching process: first match the CRD schema to the kubenertes core resource schema, then attach the generator to the matched schemas
- Rely on fetched swagger schemas with specified version from the kubernetes repo to do matching
- Use two modes (strict and heuristic) to do matching on superset and subset schemas
- Implement generators as functions that take a schema as input and return a list of testcases
- Differentiate between normal and semantic testcases by subclassing or using a class field as flag
- User-defined operator-specific generators can override the default generators by using the `generator` decorator with a higher priority which will be matched first than the default generators
- Specify the schemas each generator can handle by using the `generator` decorator

### Examples
Within acto we could annotate the generator functions with the `generator` decorator and specify the schemas each generator can handle. For example, the `gen_replicas` generator can handle the kubenertes schema `replicas` that is of type `int`. (Note there are k8s primitive schemas with multiple types) Since `replicas` is a common schema, we can implement it within acto as a default generator with a low matching priority of 0.
```python
@generator(k8s_name="replicas", k8s_format="int", priority=0)
def gen_replicas(schema):
    return [
        TestCase(
            name="k8s-replicas-scale-down",
            precondition=lambda prev: prev != None,
            mutator=lambda prev: prev + 2 if prev else 4,
            setup=lambda prev: prev - 2 if prev else 1,
        ),
        TestCase(
            name="k8s-replicas-scale-up",
            precondition=lambda prev: prev != None,
            mutator=lambda prev: prev - 2 if prev else 1,
            setup=lambda prev: prev + 2 if prev else 5,
        ),
    ]
```

For operator-specific schemas, we can use the `match` decorator with a higher priority of 1 (the default priority) to override existing generators. We can also apply function composition to reuse existing generators.
```python
# user-defined generators
@generator(predicate=lambda schema: schema.name.lower().endswith("port"))
def gen_custom_ports(schema):
    return gen_port(schema)

@generator(...)
def gen_image(schema):
    return gen_string(schema["image"], examples=['nginx', 'redis'])

# acto defined generators
@generator(k8s_name="port", k8s_format="int", priority=0)
def gen_port(schema, examples=[]):
    return [
        ...
    ]

@generator(k8s_format="string", priority=0)
def gen_string(schema, examples=[]):
    return [
        ...
    ]
```


We can also attach the generator to schemas based on paths. For example, we can attach the `gen_ports` generator to schemas with paths `/spec/ports` and `/spec/containers/ports`.
```python
@generator(paths=["/spec/ports", "/spec/containers/ports"])
def gen_ports(schema):
    return gen_port(schema["port"])
```


```python
# generator functions will be decorated with the match decorator and added to
# the generators list
generators = []

# load schemas from the kubernetes repo
kubernetes_schemas = load_schemas(version="1.16")
```

<img width="1423" alt="image" src="https://github.com/xlab-uiuc/acto/assets/46978338/23e196bf-5e8c-4362-acef-16a280a1842a">

Tyler and I agreed that it is the best to do the refactoring incrementally. I will be working on the first phase now on implementing the k8s schema matching based on the swagger file.

**Comment by tylergu** [comment]
Closed by #307
